src/funk/fd_funk_txn.c

```diff
@@ -0,0 +1,723 @@
+ulong
+fd_funk_txn_cancel_children( fd_funk_t *     funk,
+                             fd_funk_txn_t * txn,
+                             int             verbose ) {
+
+  if( FD_UNLIKELY( !funk ) ) {
+    if( FD_UNLIKELY( verbose ) ) FD_LOG_WARNING(( "NULL funk" ));
+    return 0UL;
+  }
+
+  fd_funk_txn_t * map = fd_funk_txn_map( funk, fd_funk_wksp( funk ) );
+
+  ulong txn_max = fd_funk_txn_map_key_max( map );
+
+  ulong oldest_idx;
+
+  if( FD_LIKELY( !txn ) ) oldest_idx = fd_funk_txn_idx( funk->child_head_cidx ); /* opt for non-compete */
+  else {
+
+    ulong txn_idx = (ulong)(txn - map);
+    if( FD_UNLIKELY( (txn_idx>=txn_max) /* Out of map */ | (txn!=(map+txn_idx)) /* Bad alignment */ ) ) {
+      if( FD_UNLIKELY( verbose ) ) FD_LOG_WARNING(( "txn is not a funk transaction" ));
+      return 0UL;
+    }
+
+    if( FD_UNLIKELY( !fd_funk_txn_map_query( map, &txn->id, NULL ) ) ) {
+      if( FD_UNLIKELY( verbose ) ) FD_LOG_WARNING(( "txn is not in preparation" ));
+      return 0UL;
+    }
+
+    oldest_idx = fd_funk_txn_idx( txn->child_head_cidx );
+
+  }
+  
+  return fd_funk_txn_cancel_sibling_list( funk, map, txn_max, fd_funk_txn_cycle_tag(), oldest_idx, FD_FUNK_TXN_IDX_NULL );
+}
+
+/* fd_funk_txn_publish_funk_child publishes a transaction that is known
+   to be a child of funk.  Callers have already validated our input
+   arguments.  Returns FD_FUNK_SUCCESS on success and an FD_FUNK_ERR_*
+   code on failure. */
+
+static int
+fd_funk_txn_publish_funk_child( fd_funk_t *     funk,
+                                fd_funk_txn_t * map,
+                                ulong           txn_max,
+                                ulong           tag,
+                                ulong           txn_idx ) {
+
+  /* Publish txn here.  If fail, preserve txn and return an error code. */
+
+  /* At this point, txn has been committed to permanent storage */
+
+  /* Cancel all competing transaction histories */
+
+  ulong oldest_idx = fd_funk_txn_oldest_sibling( funk, map, txn_max, txn_idx );
+  fd_funk_txn_cancel_sibling_list( funk, map, txn_max, tag, oldest_idx, txn_idx );
+
+  /* Make all the children children of funk */
+
+  ulong child_head_idx = fd_funk_txn_idx( map[ txn_idx ].child_head_cidx );
+  ulong child_tail_idx = fd_funk_txn_idx( map[ txn_idx ].child_tail_cidx );
+
+  ulong child_idx = child_head_idx;
+  while( FD_UNLIKELY( !fd_funk_txn_idx_is_null( child_idx ) ) ) { /* opt for incr pub */
+
+    ASSERT_IN_PREP ( child_idx );
+    ASSERT_UNTAGGED( child_idx );
+    map[ child_idx ].tag = tag;
+
+    map[ child_idx ].parent_cidx = fd_funk_txn_cidx( FD_FUNK_TXN_IDX_NULL );
+
+    child_idx = fd_funk_txn_idx( map[ child_idx ].sibling_next_cidx );
+  }
+
+  funk->child_head_cidx = fd_funk_txn_cidx( child_head_idx );
+  funk->child_tail_cidx = fd_funk_txn_cidx( child_tail_idx );
+
+  /* Remove the mapping */
+
+  fd_funk_txn_id_copy( funk->last_publish, &map[ txn_idx ].id );
+
+  fd_funk_txn_map_remove( map, &map[ txn_idx ].id );
+
```

Comment:
Should put a comment in the places where we expect to have to clean up transaction content. Lines 225 and 491?

Reply:
Sounds like a good idea ... I'll include it in the check in of the next layer so as to save ping-pong-ing here.

Reply:
(Actually comments are already there are 191 and line 458.)